### PR TITLE
[rsyslog/rsyslog.d]: Rsyslogd config to reduce disk writes.

### PR DESCRIPTION
**Pull request by praveen-li**
Changes:
-- Added action queue config for teamd, telemetry,
   auth, cron and syslog to reduce disk writes.

Signed-off-by: Praveen Chaudhary <pchaudhary@linkedin.com>

#### Why I did it
[rsyslog/rsyslog.d]: Rsyslogd config to reduce disk writes.

See detailed explanation and testing at:
https://github.com/praveen-li/equilibrium/blob/main/rsyslog/rsyslog_config.md

#### How I did it

-- Added action queue config for teamd, telemetry,
   auth, cron and syslog to reduce disk writes.

#### How to verify it

See detailed explanation and testing at:
https://github.com/praveen-li/equilibrium/blob/main/rsyslog/rsyslog_config.md

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

**Comment by praveen-li** [comment]
@lguohan   I think, you will like these changes. Plz let me know if any questions, Thx.
 + @qiluo-msft for review.

File: files/image_config/rsyslog/rsyslog.d/00-sonic.conf [hunk 1]
```diff
@@ -13,12 +13,12 @@ if re_match($programname, "bgp[0-9]*#bgpd") then {
 ## Teamd rules
 
 if $programname contains "teamd_" then {
-    /var/log/teamd.log
+    action(type="omfile" file="/var/log/teamd.log" flushOnTXEnd="off" asyncWriting="on" flushInterval="5" ioBufferSize="64k")
```
Comment (qiluo-msft) [inline files/image_config/rsyslog/rsyslog.d/00-sonic.conf:16]: >flushOnTXEnd [](http://example.com/codeflow?start=51&length=12)

Compare with previous config (you may know the default values), what changed?

File: files/image_config/rsyslog/rsyslog.d/99-default.conf [hunk 2]
```diff
@@ -4,9 +4,13 @@
 
 # Log all facilities to /var/log/syslog except cron, auth
 # and authpriv. They are noisy - log them to their own files
-*.*;cron,auth,authpriv.none     -/var/log/syslog
-auth,authpriv.*                 /var/log/auth.log
-cron.*                          /var/log/cron.log
+if (($syslogfacility-text == "auth") or ($syslogfacility-text == "authpriv")) then {
+        action(type="omfile" file="/var/log/auth.log" flushOnTXEnd="off" asyncWriting="on" flushInterval="5" ioBufferSize="64k")
```
Comment (qiluo-msft) [inline files/image_config/rsyslog/rsyslog.d/99-default.conf:8]: >         [](http://example.com/codeflow?start=0&length=8)

Indent 4 blanks?

File: files/image_config/rsyslog/rsyslog.d/00-sonic.conf (hunk 1 above)
Comment (praveen-li) [reply files/image_config/rsyslog/rsyslog.d/00-sonic.conf:16]: https://github.com/praveen-li/equilibrium/blob/main/rsyslog/rsyslog_config.md
Above page has details and testing. 

In summary, we are adding an action queue for each facility (say teamd), which has 64k buffer size for logs. Logs are written to SSD only 
- when buffer is full or,
- 5 secs are passed, or
- rsyslog\system shutdown.

This is done to improve SSD life, since SSD comes with fixed erase\write counts.

File: files/image_config/rsyslog/rsyslog.d/99-default.conf (hunk 2 above)
Comment (praveen-li) [reply files/image_config/rsyslog/rsyslog.d/99-default.conf:8]: Agree

**Review by qiluo-msft — APPROVED** [review]

**Comment by praveen-li** [comment]
@qiluo-msft @lguohan, this helps to increase lifespan of SSDs on sonic switch. I think, we should merge this one. Thx.

**Comment by lguohan** [comment]
@prgeor , can you review this one?

**Comment by praveen-li** [comment]
> @prgeor , can you review this one?

@lguohan @prgeor Thx for getting some attention on this. We can reduce SSD writes 10 folds using this. Description has link to testing done.

**Comment by prgeor** [comment]
@praveen-li can you elaborate how did you arrive at the buffers size=64K for 5secs interval?

**Comment by prgeor** [comment]
> @prgeor , can you review this one?

@lguohan reviewing this one

**Comment by prgeor** [comment]
@praveen-li the title suggests the disk write is reduced - do you mean the logs gets dropped when the buffer gets full? As I understand, the amount of data written over a period of time is not reduced, it just the frequency of write to syslog is reduced.

**Comment by praveen-li** [comment]
> @praveen-li can you elaborate how did you arrive at the buffers size=64K for 5secs interval?

@prgeor 
Rsyslog will wait for 1 of following 2 events before writing logs to SSD:
- max 5 secs:  i.e. if there is at least one log to be written in SSD, then rsyslog will write after 5 secs irrespective of buffer size. It is done to support debugging. We do not want that, a log is seen to debugger after a more than 5 secs or so.
- if 64k buffer is full. If the buffer is full, rsyslog will write logs in SSD before 5 secs itself. We have 5 files right now, so total 64k *5 = 320KB buffer is allocated for rsyslog. Here we can go higher if needed. For 100 logs per minutes, 64k buffer is enough to hold logs for 4-5 seconds.

In summary, there is not direct relation between 64K and 5 seconds. These are 2 separate events for rsyslog, if any one is hit first rsyslog will write in SSD.
Today for 100 logs per minutes, we write in SSD around 100 times. These many SSD writes will reduce SSD health.

**Comment by praveen-li** [comment]
>

**Comment by praveen-li** [comment]
> @praveen-li the title suggests the disk write is reduced - do you mean the logs gets dropped when the buffer gets full? As I understand, the amount of data written over a period of time is not reduced, it just the frequency of write to syslog is reduced.

@prgeor 
rsyslog never drops a log. If buffer is full, it will write in disk. Even on rsyslogd service restart or at system shutdown\reboot, rsyslog halts the shutdown and finish log writing to SSD first for remaining log in buffer.

**Review by prgeor — DISMISSED** [review]

**Comment by prgeor** [comment]
"/azp run"

**Comment by prgeor** [comment]
@lguohan could you help merge? I don't see the build failure is related to this PR

**Comment by praveen-li** [comment]
> @lguohan could you help merge? I don't see the build failure is related to this PR

@lguohan Gentle reminder to merge it. It will be a good improvement. Thx

**Comment by praveen-li** [comment]
"/azp run"

**Comment by praveen-li** [comment]
@qiluo-msft @lguohan 
Folks, hope you are doing great. I noticed that this important PR was not merge yet.
It had approval from @qiluo-msft and @prgeor.  Gentle reminder to merge it. Thanks and Cheers.

**Review by qiluo-msft — APPROVED** [review]

**Comment by praveen-li** [comment]
> Commenter does not have sufficient privileges for PR 11814 in repo sonic-net/sonic-buildimage

@qiluo-msft Seems, I do not have permission to run pipeline. It will be helpful if you can help me with permissions. Thanks again.
